mergify_engine/rules/__init__.py

```diff
@@ -75,16 +88,68 @@ async def __call__(self, obj: filter.GetAttrObjectT) -> bool:
         return self.match
 
     def get_attribute_name(self) -> str:
-        tree = self.partial_filter.tree.get("-", self.partial_filter.tree)
+        tree = typing.cast(filter.TreeT, self.partial_filter.tree)
+        tree = tree.get("-", tree)
         name = list(tree.values())[0][0]
         if name.startswith(filter.Filter.LENGTH_OPERATOR):
             return str(name[1:])
         return str(name)
 
 
+@dataclasses.dataclass
+class RuleConditionGroup:
+    condition: dataclasses.InitVar[
+        typing.Dict[
+            typing.Literal["and", "or"],
+            typing.List[typing.Union["RuleConditionGroup", RuleCondition]],
+        ]
+    ]
+    operator: typing.Literal["and", "or"] = dataclasses.field(init=False)
+    conditions: typing.List[
+        typing.Union["RuleConditionGroup", RuleCondition]
+    ] = dataclasses.field(init=False)
+    match: bool = dataclasses.field(init=False, default=False)
+    _used: bool = dataclasses.field(init=False, default=False)
+
+    def __post_init__(
+        self,
+        condition: typing.Dict[
+            typing.Literal["and", "or"],
+            typing.List[typing.Union["RuleConditionGroup", RuleCondition]],
+        ],
+    ) -> None:
+        if len(condition) != 1:
+            raise RuntimeError("Invalid condition")
+
+        self.operator, self.conditions = next(iter(condition.items()))
+        for cond in self.conditions:
+            if isinstance(cond, RuleCondition):
+                attribute = cond.get_attribute_name()
+                # TODO(sileht): remove this heuristics to allow this attribute everywhere
+                if attribute.partition("-")[0] in [
+                    "check",  # heuristic in merge/queue
+                    "status",  # heuristic in merge/queue
+                ]:
+                    raise voluptuous.Invalid(
+                        f"{attribute} cannot be used inside and/or"
+                    )
+
+    async def __call__(self, obj: filter.GetAttrObjectT) -> bool:
+        if self._used:
+            raise RuntimeError("RuleCondition cannot be re-used")
+        self._used = True
+        self.match = await filter.Filter(
+            typing.cast(filter.TreeT, {self.operator: self.conditions})
+        )(obj)
+        return self.match
+
+    def copy(self) -> "RuleConditionGroup":
+        return RuleConditionGroup({self.operator: [c.copy() for c in self.conditions]})
+
+
 @dataclasses.dataclass
 class RuleConditions:
-    conditions: typing.List[RuleCondition]
+    conditions: typing.List[typing.Union[RuleConditionGroup, RuleCondition]]
```

Comment:
Do we really need to have 3 different classes?

I could easily imagine the `RuleConditions` would be a `RuleConditionGroup` where we set it to `and` manually.

i.e:
```
conditions:
  - foo
  - bar
```
is transformed to
```
conditions:
  - and:
    - foo
    - bar
```

And you don't need 3 different classes.